lib/parsers/pdf_parser/xref_table.rb

```diff
@@ -0,0 +1,99 @@
+class FormatParser::PDFParser::XRefTable
+  include FormatParser::IOUtils
+
+  XRef = Struct.new(:idx, :offset, :generation_number, :entry_type)
+
+  def self.parse(io)
+    new(io).parse
+  end
+
+  def initialize(io)
+    @io = io
+  end
+
+  # Public: parse
+  #
+  # Returns an array of XRef objects
+  def parse
+    @io.seek(@io.size - 5)
+
+    # TODO: Sometimes the EOF ends with a line-ending
+    # return [] unless safe_read(@io, 5) == FormatParser::PDFParser::EOF_MARKER
+    return [] unless xref_table_offset
+
+    @io.seek(xref_table_offset)
+
+    parse_xref_table
+  end
+
+  private
+
+  def xref_table_offset
+    @xref_table_offset ||= begin
+      # Read the "tail" of the PDF and find the 'startxref' declaration
+      xref_table_size = 1024
+      tail_pos = xref_table_size > @io.size ? 0 : @io.size - xref_table_size
+
+      @io.seek(tail_pos)
+      tail = @io.read(xref_table_size)
+
+      # Find the "startxref" declaration and read the first group of integers after it
+      start_xref_index = tail.index('startxref')
+      return unless start_xref_index
+
+      startxref = tail.byteslice(start_xref_index, xref_table_size)[/\d+/]
+      return unless startxref
+
+      startxref.to_i
+    end
+  end
+
+  def parse_xref_table
+    xref_table = []
+    starting_idx = 0
+    num_objects = nil
+
+    while line = read_until_linebreak(@io, char_limit: 32)
+      case line
+      when /^(\d+) (\d+)$/
+        # Defines the starting number of the object and the
+        # number of objects in the table
+        starting_idx = $1.to_i
+        num_objects = $2.to_i
+      when /^(\d{10}) (\d{5}) (\w)$/
+        # The actual object offset.
+        xref_table << XRef.new(
+          starting_idx + xref_table.length,
+          $1.to_i,
+          $2.to_i,
+          $3
+        )
+      when /trailer/
+        break
+      end
+    end
+
+    # Check if the number of xrefs we got makes sense
+    if num_objects && num_objects != xref_table.length
+      raise "The xref table was declared to contain #{num_objects}"\
```

Comment:
This needs a separate exception class, otherwise it will bleed out to the outer level in `FormatParser.parse`. `.parse` swallows all the exceptions that can occur when you exceed read limits, but not the other ones. This is so that we can rapidly find bugs via Appsignal.